plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsWriter.java

```diff
@@ -0,0 +1,205 @@
+import org.apache.iceberg.puffin.BlobMetadata;
+import org.apache.iceberg.puffin.Puffin;
+import org.apache.iceberg.puffin.PuffinCompressionCodec;
+import org.apache.iceberg.puffin.PuffinReader;
+import org.apache.iceberg.puffin.PuffinWriter;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+import org.apache.iceberg.util.Pair;
+
+import javax.inject.Inject;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static com.google.common.collect.MoreCollectors.toOptional;
+import static com.google.common.collect.Streams.stream;
+import static io.trino.plugin.base.util.Closables.closeAllSuppress;
+import static io.trino.plugin.iceberg.TableStatisticsReader.APACHE_DATASKETCHES_THETA_V1_NDV_PROPERTY;
+import static java.lang.String.format;
+import static java.util.Map.Entry.comparingByKey;
+import static java.util.UUID.randomUUID;
+import static org.apache.iceberg.puffin.PuffinCompressionCodec.ZSTD;
+import static org.apache.iceberg.puffin.StandardBlobTypes.APACHE_DATASKETCHES_THETA_V1;
+
+public class TableStatisticsWriter
+{
+    private final String trinoVersion;
+
+    @Inject
+    public TableStatisticsWriter(NodeVersion nodeVersion)
+    {
+        this.trinoVersion = nodeVersion.toString();
+    }
+
+    public StatisticsFile writeStatisticsFile(
+            ConnectorSession session,
+            Table table,
+            long snapshotId,
+            Map<Integer, CompactSketch> ndvSketches)
+    {
+        Snapshot snapshot = table.snapshot(snapshotId);
+        long snapshotSequenceNumber = snapshot.sequenceNumber();
+        Schema schema = table.schemas().get(snapshot.schemaId());
+
+        Set<Integer> validFieldIds = stream(
+                Traverser.forTree((Types.NestedField nestedField) -> {
+                    Type type = nestedField.type();
+                    if (type instanceof Type.NestedType nestedType) {
+                        return nestedType.fields();
+                    }
+                    if (type instanceof Type.PrimitiveType) {
+                        return ImmutableList.of();
+                    }
+                    throw new IllegalArgumentException("Unrecognized type for field %s: %s".formatted(nestedField, type));
+                })
+                .depthFirstPreOrder(schema.columns()))
+                .map(Types.NestedField::fieldId)
+                .collect(toImmutableSet());
+
+        TableOperations operations = ((HasTableOperations) table).operations();
+        FileIO fileIO = operations.io();
+        String path = operations.locationProvider().newDataLocation(format("%s-%s.stats", session.getQueryId(), randomUUID()));
```

Comment:
The `stats` file will be created under "data" directory, not "metadata" directory. Is it intentional?

Reply:
yes, it was my intentional choice, but i am not 100% sure it's the right one.
 will move